### DISCUSSION TITLE
dual phase modelling

### DISCUSSION THREAD

**Discussion by anujyadav0693** [neperfepx/neper#647] [Q&A] [answered]
respected sir,
i want to model  a microstructure which is in dual phase like TI-AL6-V4. can you share code how can we model.
thank you.

**Comment by rquey** [comment]
I'm not well versed in Ti64 microstructures, but I know that they can be very different depending on the processing.  So, please be more specific.

**Reply by charanjo** [reply]
Hello Professor,
Will I be able to get a 2D Voronoi microstructure with just 2 grain size?

**Reply by rquey** [reply, marked as answer]
Yes. Define 2 modes for the `diameq` distribution:
```
neper -T -n 100 -morpho "diameq:0.5*lognormal(1,0.3)+0.5*lognormal(1,0.3),1-sphericity:lognormal(0.1,0.03)" -dim 2^C
neper -V n100-id1.tess -datacellcol mode -print img
```
![img](https://github.com/user-attachments/assets/0f931d56-8e89-4b8e-96d4-ba28d8183e22)

**Reply by charanjo** [reply]
Thank you professor. It works for me.

**Reply by Zdmire** [reply]
> Yes. Define 2 modes for the `diameq` distribution:
> 
> ```
> neper -T -n 100 -morpho "diameq:0.5*lognormal(1,0.3)+0.5*lognormal(1,0.3),1-sphericity:lognormal(0.1,0.03)" -dim 2^C
> neper -V n100-id1.tess -datacellcol mode -print img
> ```
> 
> ![img](https://private-user-images.githubusercontent.com/2659376/415680474-0f931d56-8e89-4b8e-96d4-ba28d8183e22.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDg1MDYzOTksIm5iZiI6MTc0ODUwNjA5OSwicGF0aCI6Ii8yNjU5Mzc2LzQxNTY4MDQ3NC0wZjkzMWQ1Ni04ZTg5LTRiOGUtOTZkNC1iYTI4ZDgxODNlMjIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDUyOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA1MjlUMDgwODE5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MjA1OGE3MTIzZDZmZTJhNWFkM2UyMmM3Y2EyYWU4NTMyNTFmMDJlODA4NjcyNjUzYjNmYTRmZGQ1NDRhZDdlOSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.DJ9Qh-HZ0_txmzJq3yz83StGjEOPnZ3oU3Y9EVhdFqE)

Hi Dr.  @rquey,
I have a question regarding the edge visualization in this model.
As shown in your code and the figure, there are two types (or modes) of grains.
I would like to highlight the edges based on the modes of their neighboring grains — that is:

Edges between grains of the same mode should be marked in one color;

Edges between grains of different modes should be marked in another color.

Could you please advise how to implement this distinction in Neper?

**Reply by rquey** [reply]
Needs version `4.10.2-16`:

```
neper -T -n 100 -morpho "diameq:0.5*lognormal(1,0.3)+0.5*lognormal(1,0.3),1-sphericity:lognormal(0.1,0.03)" -dim 2

# Built-in colors
neper -V n100-id1.tess -datacellcol mode -dataedgecol modenb+2 -print img1

# Custom colors
neper -T -loadtess n100-id1.tess -statedge modenb -for ""
awk '{if ($1==1) {print "black"} else {print "blue"}}' n100-id1.stedge > colors
neper -V n100-id1.tess -datacellcol mode -dataedgecol "file(colors)" -print img2

# Fancier
neper -V n100-id1.tess -datacellcol mode -datacelltrs 0.5 -dataedgecol "file(colors)" -dataedgerad "((domtype==1)?0.006:modenb*0.002)" -print img3
neper -V n100-id1.tess -datacellcol mode -datacelltrs 0.5 -dataedgecol "file(colors)" -dataedgerad "((domtype==1)?0.006:(modenb-1)*0.004)" -print img4
```
<img width="1200" height="900" alt="img1" src="https://github.com/user-attachments/assets/111373d6-58b7-4662-8368-971a25086de2" />
<img width="1200" height="900" alt="img2" src="https://github.com/user-attachments/assets/751842a2-ad69-461e-8c77-b833ab4a8c16" />
<img width="1200" height="900" alt="img3" src="https://github.com/user-attachments/assets/b3adf770-e14b-42e6-8588-679ef2878171" />
<img width="1200" height="900" alt="img4" src="https://github.com/user-attachments/assets/00c73ab4-c155-4791-bbac-7b79f078c0d0" />
